client/extensions/hello-dolly/state/reducer.js

```diff
@@ -0,0 +1,25 @@
+/**
+ * Internal dependencies
+ */
+import { HELLO_DOLLY_NEXT_LYRIC } from './action-types';
+import {
+	ROUTE_SET,
+	SECTION_SET,
+	SITE_SETTINGS_SAVE,
+} from 'state/action-types.js';
+
+export default function lyricIndex( state = 0, action ) {
+	switch ( action.type ) {
+		case ROUTE_SET:
+		case SECTION_SET:
+		case SITE_SETTINGS_SAVE:
+		case HELLO_DOLLY_NEXT_LYRIC:
+			return advanceToNextLyric( state );
+		default:
+			return state;
+	}
+}
+
+function advanceToNextLyric( index ) {
+	return index + 1;
```

Comment:
hearkening back to earlier comments I have made: I'm not entirely sure we've settled on the most appropriate data structure and this is one of the indicators. we have mutually dependent state in separate places that easily gets out of sync with each other